test/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/PreValidateTraderControllerSpec.scala

```diff
@@ -0,0 +1,139 @@
+import play.api.mvc.Results.NotFound
+import play.api.test.Helpers._
+import play.api.test.{FakeHeaders, FakeRequest}
+import uk.gov.hmrc.excisemovementcontrolsystemapi.connectors.PreValidateTraderConnector
+import uk.gov.hmrc.excisemovementcontrolsystemapi.fixture.{FakeAuthentication, FakeJsonParsers}
+import uk.gov.hmrc.excisemovementcontrolsystemapi.utils.TestUtils.{getPreValidateTraderErrorResponse, getPreValidateTraderRequest, getPreValidateTraderSuccessResponse}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class PreValidateTraderControllerSpec
+  extends PlaySpec
+    with FakeAuthentication
+    with FakeJsonParsers
+    with BeforeAndAfterEach {
+
+  implicit val ec: ExecutionContext = ExecutionContext.Implicits.global
+  implicit val sys: ActorSystem = ActorSystem("DraftExciseMovementControllerSpec")
+  private val connector = mock[PreValidateTraderConnector]
+  private val cc = stubControllerComponents()
+  private val request = createRequest(Json.toJson(getPreValidateTraderRequest))
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    reset(connector)
+
+    when(connector.submitMessage(any, any)(any)).
+      thenReturn(Future.successful(Right(Right(getPreValidateTraderSuccessResponse))))
+  }
+
+  "submit" should {
+    "return 200 when validated" in {
+      val result = createWithSuccessfulAuth.submit(request)
+
+      status(result) mustBe OK
+      contentAsJson(result) mustBe Json.toJson(getPreValidateTraderSuccessResponse)
+
+    }
+
+    "return 200 when business error" in {
+      when(connector.submitMessage(any, any)(any)).
+        thenReturn(Future.successful(Right(Left(getPreValidateTraderErrorResponse))))
+
+      val result = createWithSuccessfulAuth.submit(request)
+
+      status(result) mustBe OK
+      contentAsJson(result) mustBe Json.toJson(getPreValidateTraderErrorResponse)
+
+    }
+
+    "send a request to EIS" in {
+      await(createWithSuccessfulAuth.submit(request))
+
+      verify(connector).submitMessage(any, any)(any)
+
+    }
+
+    "return an error when EIS error" in {
+      when(connector.submitMessage(any, any)(any))
+        .thenReturn(Future.successful(Left(NotFound("not found"))))
+
+      val result = createWithSuccessfulAuth.submit(request)
+
+      status(result) mustBe NOT_FOUND
+    }
+
+    "a validation parser error" when {
+      "xml cannot be parsed" in {
+        val result = createWithFailingXmlParserAction.submit(request)
+
+        status(result) mustBe BAD_REQUEST
+      }
+    }
+
+    "return authentication error" when {
+      "authentication fails" in {
+        val result = createWithAuthActionFailure.submit(request)
+
+        status(result) mustBe FORBIDDEN
+      }
+    }
+
+  }
+
+  private def createWithAuthActionFailure =
+    new PreValidateTraderController(
+      FakeFailingAuthentication,
+      FakeSuccessJsonParser,
+      connector,
+      cc
+    )
+
+  private def createWithFailingXmlParserAction =
```

Comment:
Ah, relating to the previous comment, I see that createWithFailingXmlparserAction is a typo and it's actually checking failed JSON parsing

Reply:
Renamed